### PR TITLE
Refactor Logging

### PR DESCRIPTION
**Pull request by RealistikDash**
This PR aims at refactoring logging to be more useful for production environments.

## Objectives
- [x] Implement optional logz.io logging provider support.
- [x] Refactor logs to me more searchable
    - [x] Remove formattable strings from the main `message`
    - [x] Pass relevant logging information as kwargs
- [x] Increase the amount of logging (beyond just the handler level)

### CODE REVIEW

**Comment by cmyui** [comment]
Btw I wasn't particularly impressed with logz.io from my attempts - I might recommend looking at https://newrelic.com/ if you feel the same

**Comment by RealistikDash** [comment]
> Btw I wasn't particularly impressed with logz.io from my attempts - I might recommend looking at https://newrelic.com/ if you feel the same

Actually looks rather interesting, will investigate. Logz.io does feel quite clunky...

**Comment by cmyui** [comment]
> Logz.io does feel quite clunky...

this pr is def still an improvement though, looking nice

**Comment by cmyui** [comment]
it comes

**Comment by RealistikDash** [comment]
Note for future self: This currently has an issue where only logs of severity `INFO` or higher appear in logz.io. This is despite the log level being set to debug in the `.env` file. 
image

Also Uvicorn seems to bypass the custom exception hook.

File: rgdps/logger.py [hunk 1]
```diff
@@ -1,31 +1,147 @@
...
+    "formatters": {
+        "logzioFormat": {
+            "format": '{"additional_field": "value"}',
+            "validate": False,
+        },
+    },
+    "handlers": {
+        "logzio": {
+            "class": "logzio.handler.LogzioHandler",
+            "level": "INFO",
```
Comment (cmyui) [inline rgdps/logger.py]: > Note for future self: This currently has an issue where only logs of severity INFO or higher appear in logz.io. This is despite the log level being set to debug in the .env file.

it's cuz of this

**Comment by cmyui** [comment]
> Also Uvicorn seems to bypass the custom exception hook.

hmm you're right, i let it do this in my env because i thought it was a nice format for local dev & i ran gunicorn on prod..

File: rgdps/logger.py (hunk 1 above)
Comment (RealistikDash) [reply rgdps/logger.py]: Oh my bad haha

File: .env.example [hunk 2]
```diff
@@ -36,3 +36,11 @@ RGDPS_DIR=./rgdps_data
 
 # RealistikGDPS Server Configuration
 SRV_NAME=RealistikGDPS
+
+# Logging Configuration
+LOG_LEVEL=INFO
+LOGZIO_ENABLED=0
```
Comment (lenforiee) [inline .env.example:42]: make it proper bool true/false value and then parse `val == "true"` for config readibility?

File: rgdps/api/gd/user_comments.py [hunk 3]
```diff
@@ -90,13 +112,28 @@ async def like_target_post(
...
+                "target_id": target_id,
+                "is_positive": is_positive,
+                "error": result.value,
+            },
         )
         return responses.fail()
 
-    logger.info(f"{user} successfully liked {target_type!r} {target_id}.")
+    logger.info(
+        "Successfully liked target.",
```
Comment (lenforiee) [inline rgdps/api/gd/user_comments.py]: inconsistient log content, above says `like/dislike` and here only `liked`

Comment (RealistikDash) [reply rgdps/api/gd/user_comments.py]: Corrected

File: .env.example (hunk 2 above)
Comment (RealistikDash) [reply .env.example:42]: Imo, this isnt the PR for this sort of change? The current config parser would support that though.

File: rgdps/api/__init__.py [hunk 4]
```diff
@@ -205,12 +224,62 @@ def init_routers(app: FastAPI) -> None:
...
+        if str(request.url).startswith(config.http_url_prefix):
+            # GD sends an empty User-Agent header.
+            if request.headers.get("User-Agent") != "":
+                logger.info(
+                    "Client request stopped due to invalid User-Agent header.",
+                    extra={
+                        "url": str(request.url),
+                        "uuid": request.state.uuid,
+                    },
+                )
```
Comment (cmyui) [inline rgdps/api/__init__.py:254]: maybe log the user-agent

File: rgdps/api/__init__.py [hunk 5]
```diff
@@ -205,12 +224,62 @@ def init_routers(app: FastAPI) -> None:
...
+
+    @app.middleware("http")
+    async def exception_logging(
+        request: Request,
+        call_next: RequestResponseEndpoint,
+    ) -> Response:
+        try:
+            return await call_next(request)
+        except Exception as e:
+            logger.error(
```
Comment (cmyui) [inline rgdps/api/__init__.py]: use logger.exception() for cases where u wish for the exception to be included in the log

Comment (cmyui) [reply rgdps/api/__init__.py]: ^ this is the same as manually passing `exc_info=e` to `logger.error()`

File: rgdps/api/gd/user_relationships.py [hunk 6]
```diff
@@ -92,12 +115,21 @@ async def friend_request_read(
 
     if isinstance(result, ServiceError):
         logger.info(
-            f"{user} failed to mark friend request with ID {request_id} as seen with error {result!r}.",
+            f"Failed to mark friend request as seen.",
```
Comment (cmyui) [inline rgdps/api/gd/user_relationships.py]:
```suggestion
            "Failed to mark friend request as seen.",
```

File: rgdps/api/gd/user_relationships.py [hunk 7]
```diff
@@ -201,29 +276,41 @@ async def friend_remove_post(
...
         ctx,
         user.id,
         target_id,
     )
-
+    # TODO: Temp.
     if isinstance(result, ServiceError):
         logger.info(
-            f"{user} failed remove friend with ID {target_id} on block with error {result!r}.",
+            f"Failed remove friend.",
```
Comment (cmyui) [inline rgdps/api/gd/user_relationships.py:313]:
```suggestion
            "Failed remove friend.",
```

File: rgdps/logger.py [hunk 8]
```diff
@@ -1,31 +1,147 @@
...
+        },
+    },
+    "loggers": {
+        "rgdps": {
+            "level": "DEBUG",
+            "handlers": ["logzio"],
+            "propagate": True,
+        },
+    },
+}
```
Comment (cmyui) [inline rgdps/logger.py:40]: for the akatsuki prod services i've opted to have a [logging.yaml.example](https://github.com/osuAkatsuki/bancho-service/blob/main/logging.yaml.example) in the repo and have `logging.yaml` in `.gitignore`. this also allows us to customize it dynamically (e.g. different environments might have different logging configs)

File: rgdps/services/pubsub.py [hunk 9]
```diff
@@ -36,10 +40,14 @@ async def _listen_router(
                     # is a good idea.
                     handler = redis_handlers[message["channel"]]
                     await handler(ctx, message["data"])
-                except Exception:
+                except Exception as e:
                     logger.error(
```
Comment (cmyui) [inline rgdps/services/pubsub.py]: nit: probably prefer .exception over passing exc_info - same comment in a few other places

**Comment by cmyui** [comment]
generally looks good!

File: rgdps/logger.py (hunk 8 above)
Comment (RealistikDash) [reply rgdps/logger.py:40]: I will investigate this in the future as the deployment practises are made more concrete (likely with #31)

**Pull request merged**